File: 2020/day9/day9.py
#!/usr/bin/env python3
# https://adventofcode.com/2020/day/9

import sys
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()

numbers = list(map(int, data.splitlines()))

# part 1
for i in range(25, len(numbers)):
    invalid = numbers[i]
    found = False
    for a in range(i - 25, i):
        for b in range(i - 25, i):
            if numbers[a] + numbers[b] == invalid:
                found = True
                break
        if found:
            break
    if not found:
        break
print(invalid)

# part 2
weakness = 0
for i in range(len(numbers)):
    acc = 0
    for j in range(i, len(numbers)):
        acc += numbers[j]
        if acc == invalid:
            weakness = min(numbers[i : j + 1]) + max(numbers[i : j + 1])
            break
        if acc > invalid:
            break
    if weakness:
        break
print(weakness)

File: 2020/day9/day9.rs
//! [Day 9: Encoding Error](https://adventofcode.com/2020/day/9)

struct Puzzle {
    numbers: Vec<u64>,
    window: usize,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            numbers: data
                .lines()
                .map_while(|line| line.parse::<u64>().ok())
                .collect(),
            window: 25,
        }
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        'outer: for i in self.window..self.numbers.len() {
            let invalid = self.numbers[i];

            for a in &self.numbers[(i - self.window)..i] {
                for b in &self.numbers[(i - self.window)..i] {
                    if a + b == invalid {
                        continue 'outer;
                    }
                }
            }

            return invalid;
        }
        0
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        let invalid = self.part1();

        for i in 0..self.numbers.len() {
            let mut acc = 0;

            for j in i..self.numbers.len() {
                acc += self.numbers[j];
                if acc == invalid {
                    return self.numbers[i..=j].iter().min().unwrap()
                        + self.numbers[i..=j].iter().max().unwrap();
                }
                if acc > invalid {
                    break;
                }
            }
        }

        0
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u64, u64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test_part1() {
        let mut puzzle = Puzzle::new(TEST_INPUT);
        puzzle.window = 5;
        assert_eq!(puzzle.part1(), 127);
    }

    #[test]
    fn test_part2() {
        let mut puzzle = Puzzle::new(TEST_INPUT);
        puzzle.window = 5;
        assert_eq!(puzzle.part2(), 62);
    }
}
